### DISCUSSION TITLE
Unicode Error reading binary files on Arduino GIGA

### DISCUSSION THREAD

**Discussion by D-nard** [micropython/micropython#11408] [STM32 / Pyboard] [answered]
Hi, I'm trying to read non-text files using the default open mode:

```py
fin = open('file.png')
content = fin.read()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeError: 
```
I know that I should open it in binary mode, but actually this happens only on my Arduino Giga R1 (STM32H747) with ARDUINO_GIGA-20230426-v1.20.0.dfu firmware (latest release).
Trying to do the same with a Teensy 4.1 (m.iMXRT) or an ESP32S3 (Espressif DevKitC-1), it works for the purpose I'm using it (a sort of micro http server).

I did try:
```py
fin = open('file.png', mode='rb')
content = fin.read()  # the read works now
fin.close()
content.decode()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeError: 
```
Also tried decode parameter like: 'ascii','utf-8','unicode'; with same result
If I use str(content) instead of content.decode() it simply returns a doubled-backslash string:
```py
'b\'\\x89PNG\\r\\n\\x1a\\n\\x00\\x00\\x00\\rIHDR\\x00\\x00 ... 
```
If I try to get it char by char: 
```py
chr(content(i))
```

it returns this values: '\x89','P','N','G', ... which encoding is not recognized by the browser, resulting in a broken image.

Is there a way to get a byte string as string on the GIGA R1?

**Comment by dlech** [comment]
Older versions of MicroPython did not always check strings for unicode errors, so that could be the difference.

Since you say you know the data is not a string, why are you trying to turn in into a string by calling decode?

**Reply by D-nard** [reply]
Because I need to make an http (server) response containing the requested file. So I need to write the http response as string (utf-8/ascii), that's why I'm calling the decode.
To be honest I'm not sure about the encoding that is returned by default-open (text mode) and relative read, but to build a server was quite easy: only thing I needed to do is to set the right content-type (mime type 'image/png' in this case) and write the response as simple text. 
About older version, the Arduino GIGA is the latest release, but I just noticed that on Teensy and ESP32S3 (where this code works) the firmware is not updated (19.1)
I'll update both...

**Reply by dlech** [reply, marked as answer]
You can't pass a bytes object for the http response instead of a str?

**Reply by D-nard** [reply]
I did use the examples from [micropython.org - Network - TCP sockets](https://docs.micropython.org/en/latest/esp8266/tutorial/network_tcp.html) like this one:
```py
cl.send('HTTP/1.0 200 OK\r\nContent-type: text/html\r\n\r\n')
```
so I assumed that socket send was especting only strings... but I'm wrong, I'm reading it better, thanks.

**Reply by D-nard** [reply]
Changed the response into bytes, it works. Thank you!
